gpytorch/likelihoods/gaussian_likelihood.py

```diff
@@ -26,14 +30,24 @@ def _shaped_noise_covar(self, base_shape, *params):
             shape = base_shape
         return self.noise_covar(*params, shape=shape)
 
-    def forward(self, function_samples, *params, **kwargs):
-        return base_distributions.Normal(
-            function_samples, self._shaped_noise_covar(function_samples.shape, *params).diag()
-        )
+    def forward(
+        self, function_samples: Tensor, *params: Any, observation_noise: Optional[Tensor] = None, **kwargs: Any
+    ):
+        if observation_noise is None:
+            self._shaped_noise_covar(function_samples.shape, *params).diag()
+        else:
+            var = observation_noise
+        return base_distributions.Normal(function_samples, var)
 
-    def marginal(self, function_dist, *params, **kwargs):
+    def marginal(
+        self, function_dist: MultivariateNormal, *params: Any, observation_noise: Optional[Tensor] = None, **kwargs: Any
+    ) -> MultivariateNormal:
         mean, covar = function_dist.mean, function_dist.lazy_covariance_matrix
-        full_covar = covar + self._shaped_noise_covar(mean.shape, *params)
+        if observation_noise is None:
```

Comment:
Do we actually need to check this here? Won't the `noise_covar` handle this if we just pass `**kwargs` along, or am I missing something

Reply:
Yeah so that's an option, I guess it depends on whether we want to be able to pass custom observation noise also to other likelihoods. The current implementation will make sure all likelihoods derived from `_GaussianLikelihoodBase` will handle the custom observation noise. If we instead the kwargs through then we'll get null-ops from that on the other likelihoods (e.g. those with inferred noise). That seems like an inconsistency we'd like to avoid?

Reply:
I guess this could be pushed into the noise models. Unfortunately things are a little messy down there because of the different handling of inputs in `ExactGP's` `__call__` method depending on whether we're in train or eval mode, which results in things like this: https://github.com/cornellius-gp/gpytorch/blob/ebc4459d1768136db080cf23cd00faffec78e0e7/gpytorch/likelihoods/noise_models.py#L54